x/stake/keeper/keeper.go

```diff
@@ -71,21 +71,73 @@ func (k Keeper) SetPool(ctx sdk.Context, pool types.Pool) {
 	store.Set(PoolKey, b)
 }
 
-//__________________________________________________________________________
+//_______________________________________________________________________
 
-// get the current in-block validator operation counter
-func (k Keeper) InitIntraTxCounter(ctx sdk.Context) {
+// Load the last total validator power.
+func (k Keeper) GetLastTotalPower(ctx sdk.Context) (power sdk.Dec) {
 	store := ctx.KVStore(k.storeKey)
-	b := store.Get(IntraTxCounterKey)
+	b := store.Get(LastTotalPowerKey)
 	if b == nil {
-		k.SetIntraTxCounter(ctx, 0)
+		return sdk.ZeroDec()
 	}
+	k.cdc.MustUnmarshalBinary(b, &power)
+	return
 }
 
+// Set the last total validator power.
+func (k Keeper) SetLastTotalPower(ctx sdk.Context, power sdk.Dec) {
+	if !power.IsInteger() {
+		panic("input power must be whole integer")
+	}
+	store := ctx.KVStore(k.storeKey)
+	b := k.cdc.MustMarshalBinary(power)
+	store.Set(LastTotalPowerKey, b)
+}
+
+//_______________________________________________________________________
+
+// Load the last validator power.
+// Returns zero if the operator was not a validator last block.
+func (k Keeper) GetLastValidatorPower(ctx sdk.Context, operator sdk.ValAddress) (power sdk.Dec) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(GetLastValidatorPowerKey(operator))
+	if bz == nil {
+		return sdk.ZeroDec()
+	}
+	k.cdc.MustUnmarshalBinary(bz, &power)
+	return
+}
+
+func (k Keeper) powerToBytes(power sdk.Dec) []byte {
+	bz := k.cdc.MustMarshalBinary(power)
+	return bz
+}
```

Comment:
not sure this function should exist - I think we should just marshal in place

Reply:
OK, we now marshal in-place.